### ISSUE TITLE
Missing authentication after connecting to socket

### ISSUE THREAD

**Issue by JaaHann** [bitfocus/companion-module-roland-p20hd#1]
When adding this module as a new connection to companion, there are only options for the the ip-address and the polling interval.
Unfortunately I couldn't get this device to work without enabling authentication with an user-id and a password.

Adding 2 additional inputs besides the ip-address and the polling intervall to enter the user-id and the password for configuration.
After connecting to the tcp socket the user-id and the password should be send to the device inside  > socket.on('connect') {} < before > initPolling() < (index.js).

The commands for that are:
 - "stxUSR:_{user-id}_;" (responds with ACK if ok)
 - "stxPSS:_{password}_;" (responds with ACK if ok and NACK if fails)

Also "stxQIT;" should be send before closing the socket and ending communication. Otherwise connecting with a diffrent computer to the same device can be challenging, because only 1 connection via tcp is allowed.

I would love to be able to make those little changes myself, but except for knowing a tiny bit about sockets and js, I never worked with github or the companion env in general.
So I would really welcome, if someone can do those changes to make this module work properly.

Thanks :)
Jan

**Comment by josephdadams** [comment]
I can add that

**Comment by josephdadams** [comment]
Ok, I added it. Obviously, unable to test, I don't have this device. So let me know how that goes. v1.0.1 of the module, I pushed it to the beta repo but those generally don't build except once every 6 hours or so. You can download the repo and run it manually in companion v3 if you like

**Comment by JaaHann** [comment]
Thanks a lot. Just one quick note:
Even though the documentation shows ACK and NACK as possible responses for the PSS command, in general I actually never got any response at all right after sending that command. Only after requesting the version (or any other command - but VER doesn't change anything), the socket either responds with the version (correct password) or responding with NACK (incorrect password).

So if your implementation doesn't support that behavior, it would be nice if you could change that too.

Unfortunately I'm not really able to test today, but in the next days there should be some time :)

**Comment by josephdadams** [comment]
Ah ok. It relies on receiving ACK after sending password, so I’ll have to change that.

**Comment by josephdadams** [comment]
I updated the logic, please try v1.0.2 of this module and let me know how it goes

**Comment by JaaHann** [comment]
Thanks a lot, but I think this will not work because initCommunication() is only called when config.uselogin isn't true or an ACK is received. But the "PSS" command seems to responds nothing at all (or only after you send other commands, which can't be done because initCommunication() wasn't called yet).

So requesting the version should better be done right after initLoginPassword() (line 148) and not from initCommunication().
Then getting the version response and lastReturnedCommand.indexOf('PSS') > -1 should indicate a successful authentication.

Maybe something like this from line 146:
`else if (pipeline.includes(this.CONTROL_ACK)) {
    if (this.lastReturnedCommand.indexOf('USR') > -1) {
        this.initLoginPassword();
        this.sendCommand('VER'); //request version
    }
}
else if (pipeline.includes('VER:') && this.lastReturnedCommand.indexOf('PSS') > -1) {
    this.initCommunication();
}`

**Comment by josephdadams** [comment]
I think this will work:

```javascript
if (pipeline.includes(this.CONTROL_NACK)) {
	this.log('error', `Command ${this.lastReturnedCommand} was not accepted by the device`)
	if (this.lastReturnedCommand.indexOf('PSS') > -1) {
		this.log('error', `The password was probably wrong`);
	}
}
else if (pipeline.includes(this.CONTROL_ACK)) { // ACKs are sent at the end of the stream result, we should have 1 command to 1 ack
	if (this.lastReturnedCommand.indexOf('USR') > -1) { //if the last command was the username command, send the password
		this.initLoginPassword();
		this.sendCommand('VER');
	}
	else {
		//if we got an ACK at all, let's assume we're connected
		this.initCommunication();
	}
}
```

No need to check if last returned command was PSS, basically any time we get an ACK, we try to init Communication, but if we've already initCommunication, it won't do it again

**Comment by JaaHann** [comment]
But neither PSS or VER will respond with ACK.
In know the documentation is diffrent, but I never got any ACK from them.

**Comment by josephdadams** [comment]
Ok. Take a look at the latest code and give feedback, please.

**Comment by JaaHann** [comment]
This isn't working too, but I think I found the mistakes.

1. if we check for this.lastReturnedCommand.includes('USR') to send PSS, it wont be true because USR is the very first command and there is no lastReturnedCommand.
`initLoginUser() {
	this.lastReturnedCommand = 'USR'
	this.sendCommand(`USR:${this.config.username}`); //send username
}`
Just setting it to USR inside iniLiginUser solved that for me.

2. we cant really send VER because your code always waits for any response before sending the next command. But PSS will not respond right away, so VER will never be send.

Also I think I spend enough time now with your code, that I can troubbleshoot myself for a while with the device connected and then update you on the changes I made.

**Comment by JaaHann** [comment]
I found a solution for 2. but it only works when I log this.cmdPipe before and after .shift()
![grafik](https://user-images.githubusercontent.com/106606906/215276900-f414057d-5b28-4798-b095-fca856ee8d38.png)
I catch the PSS command and if there are any commands left I send the first one right after PSS.

But strangely without all the console.log() it won't work. Maby you know that behavior? Maybe some async problems...
Also please let me know, if there is a better way to communicate changes etc. I never really used Github and I don't want to make things more difficult than they need to be :)

**Comment by JaaHann** [comment]
I found some time and it works now. Besides authentication, I redid the way polling the information from the device works because the documentation of Roland is just wrong in a couple of ways (maybe the device was updated, but the documentation wasn't).
I'm now fine-tuning some actions, variables and feedbacks to get a more "streamdeck-linke" experience, especially with the new rotary actions (increasing or decreasing values instead of just setting them to a fixed value). I also asked some operators what they would love to have in terms of features the p20-hd doesn't have by default, but can be accomplished with so logic.

**Comment by josephdadams** [comment]
Can you send me that code please so I can add it here?

**Comment by JaaHann** [comment]
Sure...
https://github.com/JaaHann/companion-module-roland-p20hd.git

I would say this is about 80% finished and 60% tested.

**Comment by josephdadams** [comment]
Let me know when you feel good about it and I can pull it into the main repo and add it to the betas.

**Comment by JaaHann** [comment]
I updatet some changes and it should be completely ready now.

**Comment by josephdadams** [comment]
Great job! I updated the main repo with your updated code, and included it in the bundled-modules. It should go into the beta builds automatically sometime in the next 6 hours.
